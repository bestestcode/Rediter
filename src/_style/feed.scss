html, body {
  background-color: #fafafa;
  color: #fff;
}

body {
  height: 100vh;
  display: flex;
  flex-direction: column;
}

.rf-new-post-notification {
  display: flex;
  justify-content: center;
  margin: 15px 0px;
}

.rf-vertical-center {
  display: flex;
  align-items: center;
}

.rf-timestamp {
  margin-left: 24px;
}

a {
  text-decoration: none;
  transition: all .6s ease 0s;
  
  &:hover {
    opacity: .6;
    color: #b1afaf;
    cursor: pointer;
  }
}

.card {
  &.horizontal {
    display: flex;
    flex-grow: 1;
    flex-wrap: wrap;
  }
  
  .card-content {
    display: flex;
    flex-direction: column;
    justify-content: center;
  }

  .card-image {
    width: 160px;
    padding: 20px;
  }

  .card-expand {
    padding-left: 184px;
  }

  .card-meta {
    color: #000;
    font-weight: 300;
  }

  .card-author {
    color: #ef8361;
  }

  .card-title-text a {
    color: #000;
    font-size: 16px;
    font-weight: 400;
  }

  .card-expand {
    width: 100%;
    height: auto;
  }

  .chip-list {
    display: flex;
  }
  
  .chip {
    margin: 10px 0px 10px 12px;    
  }

  .action {
    margin-right: 12px;
    a {
      color: #ff8a65;
    }
  }
  
  i {
    &.right {
      margin-left: 4px;
    }
  }
}

.pagination {
  display: flex;
  color: #000;

  .previous {
    align-content: flex-start;
  }

  .next {
    align-content: flex-end;
  }

  .middle {
    margin: 0 auto;
  }
}

.progress-container h4 {
  color: #9ea1a8;
  font-weight: 800;
}

.img-responsive {
  max-width: 100%;
  height: auto;
  display: block;
}

.iframe-responsive iframe {
  max-width: 100%;
}

.input-field {
  color: #9e9e9e;
}

.text-right {
  text-align: right;
}

.container {
  .row {
    padding-top: 20px;

    a {
      display: inline-flex;
      padding: 0 20px;
      
      i {
        margin-right: 6px;
      }
    }
  }
}
