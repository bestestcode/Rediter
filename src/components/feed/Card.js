import React, { Component } from 'react';
import Timestamp from 'react-timestamp';

class Card extends Component {
  constructor(props) {
    super(props);
    this.state = { expanded: false };
  }

  // convert str to HTML node
  htmlDecode = (input) => {
    const e = document.createElement('div');
    e.innerHTML = input;
    return e.childNodes.length === 0 ? '' : e.childNodes[0].nodeValue;
  }

  handleExpand = () => {
    this.setState({ expanded: !this.state.expanded });
  };

  // render expand or view
  renderExpandControls() {
    const { data } = this.props.thread;
    if (data.post_hint === 'image' || data.post_hint === 'rich:video') {
      if (this.state.expanded) {
        return (
          <a className="waves-effect waves-light" onClick={this.handleExpand}>
            <i className="material-icons right">expand_less</i>Compress
          </a>
        );
      } else {
        return (
          <a className="waves-effect waves-light" onClick={this.handleExpand}>
            <i className="material-icons right">expand_more</i>Expand
          </a>
        );
      }
    } else {
      return (
        <a className="waves-effect waves-light" href={data.url} target="_blank">
          <i className="material-icons right">visibility</i>View
        </a>
      );
    }
  }

  // render Expanded Content - either image or iframe video.
  renderExpandContent() {
    const { data } = this.props.thread;

    if (data.post_hint === 'image') {
      return <img src={data.url} className="img-responsive" alt={data.title} />;
    } else if (data.post_hint === 'rich:video') {
      return <div className="iframe-responsive" dangerouslySetInnerHTML={{ __html: this.htmlDecode(data.secure_media_embed.content) }} />;
    }
  }

  render() {
    const { data } = this.props.thread;

    // Show placeholder for non-image(string) thumbnails
    switch (data.thumbnail) {
      case 'self':
      case 'nsfw':
      case 'spoiler':
      case 'default':
      case '':
        data.thumbnail = './assets/images/placeholder.png';
        break;

      default:
        break;
    }

    return (
      <div key={data.id} className="col s12">
        <div className="card hoverable horizontal">
          <div className="card-image rf-vertical-center">
            <img src={data.thumbnail} alt={data.title} />
          </div>

          <div className="card-stacked">
            <div className="card-content">
              <span className="card-title-text">
                <a href={`http://reddit.com${data.permalink}`} target="_blank" rel="noopener">
                  {data.title}
                </a>
              </span>

              <div className="card-meta">
                Posted by <a
                  className="card-author"
                  href={`https://www.reddit.com/user/${data.author}`}
                  target="_blank"
                  rel="noopener"
                >{data.author}</a>
                <Timestamp className="rf-timestamp" time={data.created_utc} />
              </div>

              <div className="chip-list">
                <span className="action rf-vertical-center">
                  {this.renderExpandControls()}
                </span>
                <div className="chip">/r/{data.subreddit}</div>
                {data.over_18 && <div className="chip">nsfw</div>}
                {data.post_hint && <div className="chip">{data.post_hint == 'rich:video' ? 'video' : data.post_hint}</div>}
              </div>
            </div>
          </div>

          {this.state.expanded &&
          <div className="card-expand">
            {this.renderExpandContent()}
          </div>}
        </div>
      </div>
    );
  }
}

export default Card;
