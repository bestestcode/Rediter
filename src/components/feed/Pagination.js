import React, { Component } from 'react';
import { Link, withRouter } from 'react-router-dom';

class Pagination extends Component {
  constructor(props) {
    super(props);

    this.state = {
      currentPage: 1
    };
  }

  // Resets the page number when the user changes route.
  componentWillReceiveProps(nextProps) {
    if (nextProps.sub !== this.props.sub) {
      this.setState({ currentPage: 1 });
    }
  }

  nextPage = () => {
    this.setState({ currentPage: this.state.currentPage + 1 });
  }

  prevPage = () => {
    this.setState({ currentPage: this.state.currentPage - 1 });
  }

  formatPathForBefore() {
    const { location } = this.props;
    return `${location.pathname}?before=t3_`;
  }

  formatPathForAfter() {
    const { location } = this.props;
    return `${location.pathname}?after=t3_`;
  }

  render() {
    return (
      <ul className="pagination col s12">
        {this.state.currentPage > 1 &&
        <li className="previous">
          <Link to={`${this.formatPathForBefore()}${this.props.firstItem}`} onClick={this.prevPage}>
            <i className="material-icons">chevron_left</i>
          </Link>
        </li>}

        <li className="middle">Page {this.state.currentPage}</li>

        <li className="next">
          <Link to={`${this.formatPathForAfter()}${this.props.lastItem}`} onClick={this.nextPage}>
            <i className="material-icons">chevron_right</i>
          </Link>
        </li>
      </ul>
    );
  }
}

export default withRouter(Pagination);
