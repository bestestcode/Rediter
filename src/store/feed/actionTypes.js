export const FETCH_THREAD_LIST_REQUEST = 'FETCH_THREAD_LIST_REQUEST';
export const FETCH_THREAD_LIST_SUCCESS = 'FETCH_THREAD_LIST_SUCCESS';
export const FETCH_THREAD_LIST_ERROR = 'FETCH_THREAD_LIST_ERROR';
export const NOTIFICATION_THREAD_NEW = 'NOTIFICATION_THREAD_NEW';
