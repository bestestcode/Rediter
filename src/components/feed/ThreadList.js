import React, { Component } from 'react';

import Card from './Card';
import Pagination from './Pagination';

export class ThreadList extends Component {
  constructor(props) {
    super(props);

    this.state = {
      firstItem: '',
      lastItem: ''
    };
  }

  // Assigns the 1st and last item in the array and binds it to state.
  componentWillReceiveProps(nextProps) {
    const { list } = nextProps.threads;
    if (list && list.length) {
      const lastItem = list.slice(-1).pop().data.id;
      const firstItem = list[0].data.id;

      if (this.state.firstItem !== firstItem || this.state.lastItem !== lastItem) {
        this.setState({
          firstItem,
          lastItem
        });
      }
    }
  }

  // maps threads to the props of Card component
  renderThread() {
    const { list } = this.props.threads;
    if (list && list.length) {
      return list.map((thread, index) => (
        <Card key={thread.data.id} thread={thread} />
      ));
    } else {
      return (
        <div className="progress-container col s12">
          {this.renderError()}
        </div>
      );
    }
  }

  // Show toast or progress
  renderError() {
    const { toast } = this.props;
    if (toast.length) {
      return <div><h4>{toast}</h4></div>;
    } else {
      return (
        <div className="progress">
          <div className="indeterminate" />
        </div>
      );
    }
  }

  render() {
    return (
      <div>
        {this.renderThread()}

        {!this.props.toast &&
        <Pagination
          {...this.props}
          lastItem={this.state.lastItem}
          firstItem={this.state.firstItem}
          location={this.props.location}
          params={this.props.params}
        />}
      </div>
    );
  }
}

export default ThreadList;
