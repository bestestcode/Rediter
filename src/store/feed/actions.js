import axios from 'axios';
import * as actionTypes from './actionTypes';
import * as redditService from './service';

export const fetchList = (subreddit, search) => (dispatch) => {
  let path = '';

  // Chages the path based on the query
  if (subreddit === null) {
    path = '/new/.json';
  } else {
    path = `/r/${subreddit}/new/.json`;
  }

  if (search) {
    path = `${path}${search}`;
  }

  dispatch(request());

  return redditService.getThreads(path)
    .then((res) => {
      dispatch(success(res));
    })

    .catch((error) => {
      dispatch(failure(error));
    });

  function request() { return { type: actionTypes.FETCH_THREAD_LIST_REQUEST }; }
  function success(threads) { return { type: actionTypes.FETCH_THREAD_LIST_SUCCESS, payload: threads }; }
  function failure(error) { return { type: actionTypes.FETCH_THREAD_LIST_ERROR, payload: { response: error, invalid: true } }; }
};
