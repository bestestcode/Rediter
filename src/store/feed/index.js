import * as feedActions from './actions';

export { feedActions };
