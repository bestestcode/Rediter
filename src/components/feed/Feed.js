import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { feedActions } from '../../store/feed';
import { subredditsActions } from '../../store/subreddits';
import ThreadList from './ThreadList';

export class Feed extends Component {
  componentWillReceiveProps(nextProps) {
    if (nextProps.location.search !== this.props.location.search) {
      this.fetchContent(nextProps.location.search);
    }
  }

  // call action creator to fetch feed
  fetchContent = (search) => {
    const { fetchList, subreddits } = this.props;

    const subs = subreddits
      .filter(item => item.favorite)
      .reduce((str, item) => `${str}+${item.title}`, '');
    if (search) {
      fetchList(subs, search);
    } else {
      fetchList(subs, this.props.location.search);
    }
  }

  refreshContent = () => {
    this.fetchContent(null);
  }

  render() {
    const { newPosts } = this.props.threads;
    return (
      <div>
        {newPosts > 0 &&
          <div className="rf-new-post-notification">
            <button className="btn" onClick={this.refreshContent}>
              New Threads Posted
            </button>
          </div>}
        <ThreadList {...this.props} />
      </div>
    );
  }
}

Feed.contextTypes = {
  router: PropTypes.object
};

export default connect(
  state => ({
    threads: state.threads,
    toast: state.threads.toast,
    subreddits: state.subreddits.items
  }),

  dispatch => bindActionCreators({ ...feedActions, ...subredditsActions }, dispatch),
)(Feed);
