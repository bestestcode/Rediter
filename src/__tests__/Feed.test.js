import React from 'react';
import { mount, shallow } from 'enzyme';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';

import { Feed } from '../components/feed/Feed';
import { ThreadList } from '../components/feed/ThreadList';

describe('-------------COMPONENT TESTING - Feed ------------- ', () => {
  const wrapper = shallow(<Feed />);

  it('+++ Feed Dumb Component Rendering', () => {
    expect(wrapper.length).toEqual(1);
  });

  it('Feed function snapshot testing', () => {
    // const renderedResult = renderer.create(<Feed />).toJSON;
    expect(wrapper).toMatchSnapshot();
  });

  it('+++ Feed Contains ThreadList Component', () => {
    expect(wrapper.contains(<ThreadList />)).toBe(true);
  });
});
