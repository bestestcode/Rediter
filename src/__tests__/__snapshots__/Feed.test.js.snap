// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`-------------COMPONENT TESTING - Feed -------------  Feed function snapshot testing 1`] = `<ThreadList />`;
